### PR TITLE
Feature/sdk 833 | Location & Auto-complete

### PR DESCRIPTION
**Pull request by nicolas-chaix**
Fixes: SDK-833 & SDK-836
Editing readme.md -> transport mode and form usage

Warning: use the coverage "fr-idf" for validating distance info ("stif" doesn't have distance info)

### CODE REVIEW

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/domain/AutoCompletionDomain.java [hunk 1]
```diff
@@ -10,7 +10,7 @@
     private String id = "";
     private AutoCompletionType type = AutoCompletionType.ADDRESS;
     private String name = "";
-    private int distance = -1;
+    private String distance = "-1";
```
Comment (Guitouille) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/domain/AutoCompletionDomain.java]: An empty string wouldn't be better ?

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java [hunk 2]
```diff
@@ -262,6 +340,10 @@ private void showError(String message) {
 
     @Override
     public void onAutoCompletionItemClick(String id, String name, String type) {
+
+        if ("-1".equals(id))
```
Comment (Guitouille) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java]: I don't think having an hardcoded value in the code is good. I would rather we tests if the value is empty.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/journeys/JourneysFragment.java [hunk 3]
```diff
@@ -214,6 +219,17 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
         return fragmentView;
     }
 
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+
+        if (ActivityCompat.checkSelfPermission(view.getContext(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
```
Comment (Guitouille) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/journeys/JourneysFragment.java]: Maybe create a helper for the check permission, since it is present in different place in the code ?

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/model/AutoCompletionModel.java [hunk 4]
```diff
@@ -13,7 +13,7 @@
     private AutoCompletionType type = AutoCompletionType.ADDRESS;
     private Drawable icon = null;
     private String name = "";
-    private int distance = -1;
+    private String distance = "-1";
```
Comment (Guitouille) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/model/AutoCompletionModel.java]: Same as first comment.

File: navitiasdkux/src/test/java/org/kisio/navitia/sdk/ui/util/DistanceUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+import static org.junit.Assert.*;
+@RunWith(RobolectricTestRunner.class)
+@Config(qualifiers = "en")
+public class DistanceUtilTest {
+
+    @Test
+    public void getDistance_meter_zero() {
```
Comment (Guitouille) [inline navitiasdkux/src/test/java/org/kisio/navitia/sdk/ui/util/DistanceUtilTest.java]: Maybe add a test case for the null value ? (juste in case)

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java (hunk 2 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java]: done

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/model/AutoCompletionModel.java (hunk 4 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/model/AutoCompletionModel.java]: done

File: navitiasdkux/src/test/java/org/kisio/navitia/sdk/ui/util/DistanceUtilTest.java (hunk 5 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/test/java/org/kisio/navitia/sdk/ui/util/DistanceUtilTest.java]: done for null and empty string

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/journeys/JourneysFragment.java (hunk 3 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/journeys/JourneysFragment.java]: done and added some annotations in order to remove Missing Permissions Warning from Lint

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java [hunk 6]
```diff
@@ -120,10 +135,43 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
 
     @Override
     public void onStop() {
-        super.onStop();
         if (getActivity() != null && getActivity().getWindow().getCurrentFocus() != null) {
             Helper.hideKeyboard(getActivity(), getActivity().getWindow().getCurrentFocus().getWindowToken());
         }
+
+        if (timer != null)
+            timer.cancel();
```
Comment (EvenModiguy) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java]: Please add brackets if this isn't on the same line.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java [hunk 7]
```diff
@@ -188,8 +236,17 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
 
             @Override
             public void afterTextChanged(Editable s) {
+                if (timer != null)
+                    timer.cancel();
```
Comment (EvenModiguy) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java]: Please add brackets if this isn't on the same line.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java [hunk 8]
```diff
@@ -28,31 +29,48 @@
...
+    public void executeRequest(String coords) {
+        getAddressFromPosition.execute(new AddressFromPositionObserver(), GetAddressFromPosition.Params.forTypeCoord(coords));
+    }
+
     void reverseGeoCoding(String originCoords, String destinationCoords) {
-        List<Pair<Integer, String>> coordsList = new ArrayList<>(2);
+        List<Pair<Integer, String>> coordsList = new ArrayList<>();
         coordsList.add(new Pair<>(0, originCoords));
-        coordsList.add(new Pair<>(1, destinationCoords));
+        if (!destinationCoords.isEmpty())
```
Comment (EvenModiguy) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java]: Please add brackets if this isn't on the same line.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java [hunk 9]
```diff
@@ -88,19 +106,47 @@ public void onSubscribe(Disposable d) {
...
-            List<AutoCompletionModel> autoCompletionModelList = autoCompletionModelDataMapper.transform(autoCompletionDomainList);
 
-            AutoCompletionModel myPosition = new AutoCompletionModel();
+            final List<AutoCompletionModel> autoCompletionModelList = autoCompletionModelDataMapper.transform(autoCompletionDomainList);
+
+            lastAutoCompletionModelList.clear();
+            lastAutoCompletionModelList.addAll(autoCompletionModelList);
+
+            if (myPosition != null && isDepartureMode)
+                autoCompletionModelList.add(0, myPosition);
```
Comment (EvenModiguy) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java]: Please add brackets if this isn't on the same line.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java [hunk 10]
```diff
@@ -88,19 +106,47 @@ public void onSubscribe(Disposable d) {
...
+
+    private final class AddressFromPositionObserver implements SingleObserver<AddressDomain> {
+        @Override
+        public void onSubscribe(Disposable d) {
+
+        }
+
+        @Override
+        public void onSuccess(AddressDomain addressDomain) {
+
```
Comment (EvenModiguy) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java]: Please remove this empty line.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/util/DistanceUtil.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
+package org.kisio.navitia.sdk.ui.util;
+
+import android.text.TextUtils;
+
+import java.util.Locale;
+
+public class DistanceUtil {
+
```
Comment (EvenModiguy) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/util/DistanceUtil.java]: Please remove this empty line.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/util/DistanceUtil.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    private static final String DISTANCE_METER_PATTERN = "%d m";
+    private static final String DISTANCE_KILOMETER_PATTERN = "%.2f km";
+
+    public static String getDistance(String measureString) {
+
```
Comment (EvenModiguy) [inline navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/util/DistanceUtil.java]: Please remove this empty line.

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java (hunk 6 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java]: done

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java (hunk 7 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteFragment.java]: done

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java (hunk 8 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/presentation/autocomplete/AutoCompleteViewModel.java]: done

File: navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/util/DistanceUtil.java (hunk 12 above)
Comment (nicolas-chaix) [reply navitiasdkux/src/main/java/org/kisio/navitia/sdk/ui/util/DistanceUtil.java]: done
